larpandoradlcontent/LArTwoDReco/DlTrackShowerStreamSelectionAlgorithm.cc

```diff
@@ -0,0 +1,75 @@
+/**
+ *  @file   larpandoradlcontent/LArTwoDReco/DlTrackShowerStreamSelectionAlgorithm.cc
+ *
+ *  @brief  Implementation of the deep learning track shower cluster streaming algorithm.
+ *
+ *  $Log: $
+ */
+
+#include "Pandora/AlgorithmHeaders.h"
+
+#include "larpandoradlcontent/LArTwoDReco/DlTrackShowerStreamSelectionAlgorithm.h"
+
+#include "larpandoracontent/LArHelpers/LArMCParticleHelper.h"
+#include "larpandoracontent/LArHelpers/LArMonitoringHelper.h"
+
+#include "larpandoracontent/LArObjects/LArCaloHit.h"
+
+#include <numeric>
+
+using namespace pandora;
+using namespace lar_content;
+
+namespace lar_dl_content
+{
+
+StatusCode DlTrackShowerStreamSelectionAlgorithm::AllocateToStreams(const Cluster *const pCluster)
+{
+    const OrderedCaloHitList &orderedCaloHitList{pCluster->GetOrderedCaloHitList()};
+    CaloHitList caloHits;
+    orderedCaloHitList.FillCaloHitList(caloHits);
+    const CaloHitList &isolatedHits{pCluster->GetIsolatedCaloHitList()};
+    caloHits.insert(caloHits.end(), isolatedHits.begin(), isolatedHits.end());
+    FloatVector trackLikelihoods;
+    try
+    {
+        for (const CaloHit *pCaloHit : caloHits)
+        {
+            const LArCaloHit *pLArCaloHit{dynamic_cast<const LArCaloHit *>(pCaloHit)};
+            const float pTrack{pLArCaloHit->GetTrackProbability()};
+            const float pShower{pLArCaloHit->GetShowerProbability()};
+            if ((pTrack + pShower) > std::numeric_limits<float>::epsilon())
+                trackLikelihoods.emplace_back(pTrack / (pTrack + pShower));
+        }
+
+        const unsigned long N{trackLikelihoods.size()};
+        if (N > 0)
+        {
+            float mean{std::accumulate(std::begin(trackLikelihoods), std::end(trackLikelihoods), 0.f) / N};
+            if (mean >= 0.5f)
+                m_clusterListMap.at(m_trackListName).emplace_back(pCluster);
+            else
+                m_clusterListMap.at(m_showerListName).emplace_back(pCluster);
+        }
+    }
```

Comment:
This code is very similar to the code in `DlPfoCharacterisationAlgorithm.cc`... could the clear track function be made generic in a helper and used here?